### PR TITLE
#Prepare codebase for TSE review

### PR DESCRIPTION
**Pull request by codecell**
**What does this PR do?**
Gets the codebase ready for TSE review

**How can this manually tested?**
- After cloning this repo to your favorite location
- `cd ruby enumerables`
- in your terminal run `ruby test.rb` to see the result of all the test cases covered

### CODE REVIEW

**Review by whiz25 — CHANGES_REQUESTED** [review]
## Hey @codecell 
I know you have worked hard in this project but you haven't followed the correct git-flow and therefore this PR will be marked as invalid.
- [ ] Correct Git Flow: production `master` branch + main `development` branch + Pull Requests from feature branches to the `develop` ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/git_flow.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/57421589/))

#### [FIX]
- [ ] create a feature branch and make the PR from feature to develop.
- [ ] fix the stickler errors
- [ ] fix the conflicts

### Martin Afani

**Comment by codecell** [comment]
Hi @whiz25, Thanks for corrections, I've implemented the requested changes
- deleted all the other redundant branches
- created a new `development` branch off of master
- created a new `feature/ruby-enumerables` branch
- I did not delete this `develop` branch since this PR is created from it
- the new  `feature/ruby-enumerables` is exactly the same as the `develop` branch
- once this PR is approved, I will merge `feature/ruby-enumerables` to `development` and delete this `develop` branch

**Review by whiz25 — CHANGES_REQUESTED** [review]
## Nice Progress :+1: 
You have made great strides with the project given that it's your first attempt at the project. However, we can not merge this project yet since we have some failing edge cases to fix.
Check my comments for suggestions.
### [ Things to note ]
- [ ] follow the right git-flow in your next projects
- [ ] make your PR more descriptive. Tell us what features are covered by the PR

### Martin Afani

File: enumerables.rb
Comment (whiz25) [inline enumerables.rb]: - [ ] should be identical and returns the same thing as ruby's [`each`](https://ruby-doc.org/core-2.6.4/Array.html#method-i-each). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [ ] my_each returns an Enumerator if no block is given
```ruby
array = Array.new(100) { rand(0...9) } }
array.my_each) == Enumerator #1 true
```

File: enumerables.rb
Comment (whiz25) [inline enumerables.rb]: - [ ] should be identical and returns the same thing as ruby's [`each_with_index`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-each_with_index). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each_with_index.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [ ] returns an enumerator if no block is given
```ruby 
      my_each_output = ''
      block = proc { |num, idx| my_each_output += "Num: #{num}, idx: #{idx}\n" }
      array.each_with_index(&block)
      each_output = my_each_output.dup
      my_each_output = ''
      array.my_each_with_index(&block)
      my_each_output == each_output #true
```

File: enumerables.rb
Comment (whiz25) [inline enumerables.rb]: - [ ] should be identical and returns the same thing as ruby's [`select`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-select). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_select.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [ ] returns an enumerator if no block is given
```ruby
array.my_select == Enumerator #true
```
- [ ] returns an array containing all elements of enum for which the given block returns a true value
```ruby
array = Array.new(100){rand(0...9)}
array.my_select(&block) == array.select(&block) #true
```
```

File: enumerables.rb
Comment (whiz25) [inline enumerables.rb]: - [ ] should be identical and returns the same thing as ruby's [`all?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_all.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [ ] when no block or argument is given returns true when none of the collection members are false or nil
```ruby
true_array = [1, true, 'hi', []] 
false_array = [1, false, 'hi', []] 
true_array.my_all? == true_array.all? #true
```
- [ ] when a class is passed as an argument returns true if all of the collection is a member of such class
```ruby
array = Array.new(100){rand(0...9)}
array.my_all?(Integer) == array.all?(Integer)
```
- [ ] when a Regex is passed as an argument returns true if all of the collection matches the Regex
```ruby
words = %w[dog door rod blade]
words.my_all?(/d/) == words.all?(/d/)
```
- [ ] when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern
```ruby
array.my_all?(3) == array.all?(3) #true
```

File: enumerables.rb
Comment (whiz25) [inline enumerables.rb]: - [ ] should be identical and returns the same thing as ruby's [`any?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_any.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [ ] when no block or argument is given returns true if at least one of the collection is not false or nil
```ruby
true_array = [nil, false, true, []]
true_array.my_any? == true_array.any? #true
```
- [ ]  when a class is passed as an argument returns true if at least one of the collection is a member of such class
```ruby
words = %w[dog door rod blade]
words.my_any?(Integer) == words.any?(Integer) #true
```
- [ ] when a Regex is passed as an argument returns false if none of the collection matches the Regex
```ruby
words.my_any?(/z/) == words.any?(/z/) #true
```
- [ ] when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern
```ruby
words.my_any?('cat') == words.any?('cat') #true
```

File: enumerables.rb
Comment (whiz25) [inline enumerables.rb]: - [ ] should be identical and returns the same thing as ruby's [`none?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_none.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [ ] when a class is passed as an argument returns true if none of the collection is a member of such class
```ruby
array = Array.new(100) { rand(0...9) }
array.my_none?(String) == true #true
```
- [ ] when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern
```ruby
words = %w[dog door rod blade]
words.my_none?(5) == words.none?(5) #true
```
- [ ] Regex is passed as an argument returns true only if none of the collection matches the Regex
```ruby
 words.my_none?(/z/) == words.none?(/z/) #true
 words.my_none?(/d/) == words.none?(/d/) #true
```
- [ ] when no block or argument is given
```ruby
true_array = [nil, false, true, []]
false_array = [nil, false, nil, false]
false_array.my_none? #true
true_array.my_none? #false
```

File: enumerables.rb
Comment (whiz25) [inline enumerables.rb]: - [ ] should be identical and returns the same thing as ruby's [`count`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-count). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_count.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [ ] returns the number of items in enum through enumeration
```ruby
array = Array.new(100){rand(0...9)}
array.my_count == array.count #true
```
- [ ] counts the number of items in enum that are equal to item if an argument is given
```ruby
array.my_count(0) == array.count(0) #true
```

File: enumerables.rb
Comment (whiz25) [inline enumerables.rb]: - [ ] should be identical and returns the same thing as ruby's [`map`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-map). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_map.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [ ] returns a new array with the results of running block once for every element in enum.
```ruby
array = Array.new(100){rand(0...9)}
array.my_map(&block) == array.map(&block) #true
```
- [ ] returns an Enumerator if no block is given
```ruby
array.my_map == Enumerator #true
```

File: enumerables.rb
Comment (whiz25) [inline enumerables.rb]: - [ ] should be identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [ ] when a symbol is specified combines each element of the collection by applying the symbol as a named method
```ruby
array = Array.new(100) { rand(0...9) }
array.my_inject(:+) == array.inject(:+) #true
```

File: enumerables.rb
Comment (codecell) [reply enumerables.rb]: Hi @whiz25, I don't understand what exactly you want me to do here, because i evn tested it with normal `each` method and got a `a false`
image

File: enumerables.rb
Comment (codecell) [reply enumerables.rb]: Hi @whiz25, please endeavor to run these cases in your terminal before your next feedback,
just take a look at this screenshot you'd see that 
`array.my_select == Enumerator #true` actually is `false`

image

File: enumerables.rb
Comment (bolah2009) [inline enumerables.rb]: What @whiz25 was saying here is that the `my_each` method should return an `Enumerator` when no block is given. Check if a block is given at the beginning of the method if no block is given return an `Enumerator` instead else continue with the method logic.

See the difference below:

```ruby
[1,2,3].each
#=> #<Enumerator: [1, 2, 3]:each>

[1,2,3].my_each
#=> [1, 2, 3]
```

HINT: Add `return self.to_enum unless block_given?`
See suggestion
```suggestion
  def my_each
    return self.to_enum unless block_given?
    
    size = count - 1
    counter = 0
    items = to_a
    while counter <= size
      if is_a? Range
        items.length >= 0 && yield(items[counter])
      elsif is_a? Hash
        yield(items[counter][0], items[counter][1])
      elsif !block_given?
        return self
      else
        yield(self[counter])
      end
      counter += 1
    end
    self
  end
```


### Let's do this for `my_each_with_index` and `my_select`

File: enumerables.rb
Comment (bolah2009) [inline enumerables.rb]: Could you test with ruby's `select`? This method should return an array contained in the initial enumerable that satify the condition in the block given. See the example below.

```ruby
[1,2,3,7,8,9].select{|i| i <  5}
#=> [1, 2, 3]

[1,2,3,7,8,9].my_select{|i| i <  5}
#=> [true, true, true, false, false, false]
```

HINT

```suggestion
  def my_select
    return self.to_enum unless block_given?
    
    required = []

    my_each do |r|
      required.push(r) if yield(r)
    end
    required
  end
```

File: enumerables.rb
Comment (bolah2009) [inline enumerables.rb]: #### It is important to note that `my_any?`, `my_all?` and `my_none?` will always return `true` or `false`

 - When a pattern is passed this method should return true if every collection member matches this pattern. Note that pattern could be a `Class` (e.g. `Integer`, `String`) or `RegEx`
   ```ruby
   # When a Class pattern is passed
   irb(main):014:0> [1,2,3].all?(Integer)
   => true
   irb(main):015:0> [1,2,3].my_all?(Integer)
   => false

   # When a RegEx pattern is passed
   irb(main):018:0> %w[ant bat cat].all?(/t/)
   => true
   irb(main):019:0> %w[ant bat cat].my_all?(/t/)
   => false
   ```
   Hint: Use ` === ` to check if a member matches a patern.

### Find more details about [`all?` here](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F)

### Please let's do this for `my_any?` and `my_none?` too.

**Review by bolah2009 — CHANGES_REQUESTED** [review]
Hi @codecell 

Good job with the project so far, however, there are still some issue with some methods as pointed out by @whiz25.

See my comments for more details, you should also check [the ruby doc for these methods behaviour](https://ruby-doc.org/core-2.6.5/Enumerable.html). Methods to look out for are the following:

- `my_each`
- `my_each_with_index`
- `my_select`
- `my_all?`
- `my_any?`
- `my_none?`
- `my_map`
- `my_inject`

#### :recycle:  Make these changes and submit another code review request. 👍🏾

#### Feel free to reach me on slack if you having issues implementing these features. :relaxed: 

Good luck with your next submission. :star2: 

Cheers
[Bola](https://bolabuari.com/)

###### PS: Note that comments tagged as `OPTIONAL` are not part of the minimum requirements to approve this PR

**Review by kobiyoyo — CHANGES_REQUESTED** [review]
## Hi @codecell       
#### Great Job so far, but there are a few more things required. 
###### Kindly see my comments for more details.
- [ ] `my_map`  
- [ ] `my_all?`
- [ ] `my_any`
- [ ] `my_none`
- [ ] `my_inject`        
### Make changes and submit another code review request.
*Happy coding*
(Any questions Feel free to contact me on slack @kobiyoyo)

File: enumerables.rb
Comment (kobiyoyo) [inline enumerables.rb]: - For `my_map` returns an Enumerator if no block is given:
`p [1, 2, 3].my_map should return <Enumerator>`
-  `my_map` returns a new array with the results of running block once for every element in enum.
##### For example
`block =  proc { |num| num < (4 + 3) / 2 }`
`p [1,22,34,5].my_map(&block) ==  [true]`
`p [1,22,34,5].map(&block) == [true, false, false, false]`

File: enumerables.rb
Comment (kobiyoyo) [inline enumerables.rb]: - For `my_inject` when a symbol is specified combines each element of the collection by applying the symbol as a named method.
##### For example
`p [2,3,3,4].my_inject(:+) == [2, 3, 3, 4]`
`p [2,3,3,4].inject(:+) == [12]`

File: enumerables.rb
Comment (kobiyoyo) [inline enumerables.rb]: - For `my_none` when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern
#### For example
`p ['cat','dog'].my_none?('fish') == false`
`p ['cat','dog'].none?('fish') == true`

File: enumerables.rb
Comment (kobiyoyo) [inline enumerables.rb]: - For `my_any` when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern
##### For example
`p ['cat','dog'].my_any?('fish') == true`
`p ['cat','dog'].any?('fish') == false`

File: enumerables.rb
Comment (kobiyoyo) [inline enumerables.rb]: - For `my_all?` when a Regex is passed as an argument returns true if all of the collection matches the Regex
#### For example
`p ['hello','chubi'].my_all?(/g/) == true` 
`p ['hello','chubi'].all?(/g/) == false`
- when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern
##### For example
` p [3,3,3,3,3,3].my_all?(3) == [3,3,3,3,3,3].all?(3)`  should return `true`

**Review by kobiyoyo — CHANGES_REQUESTED** [review]
## Hi @codecell         
#### Great Job so far,you are almost there. 
 - [ ]  `my_inject`
- For when a symbol is specified combines each element of the collection by applying the symbol as a named method.
##### For example
`p (1..2).my_inject(2, :*) == (1..2).inject(2, :*)` should return `true` 
- [ ]  `my_map`
-  it returns a new array with the results of running block once for every element in enum.
##### For example
`block =  proc { |num| num < (4 + 3) / 2 }`
`p [1,22,34,5].my_map(&block) ==  [true]`
`p [1,22,34,5].map(&block) == [true, false, false, false]`
### Suggestions
- [my_inject](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject)
- [my_map](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-map)
### Make changes and submit another code review request.
*Happy coding* 
(Any questions Feel free to contact me on slack @kobiyoyo)

File: enumerables.rb
Comment (suretrust) [inline enumerables.rb]: * [ ] should have working `my_inject` identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
```ruby
(5..10).my_inject(4) { |prod, n| prod * n } != (5..10).inject(4) { |prod, n| prod * n }
```

**Review by suretrust — CHANGES_REQUESTED** [review]
Hi @codecell, 👋 

### Great job implementing the suggested changes. 🙌 Just one more test case to go. 👇 

**Kindly submit for another code review request once this has been fixed.** _Happy coding!_ 🎉

**Comment by Anna-Myzukina** [comment]
## Hi Ezaka Alfred ! Your project looks good :+1:
# Status: [APPROVED]


### Go ahead and merge to master.


#### Goodluck in your next project Happy coding <g-emoji class="g-emoji" alias="computer" fallback-src="https://github.githubassets.com/images/icons/emoji/unicode/1f4bb.png">computer</g-emoji>

**Review by Anna-Myzukina — APPROVED** [review]

**Pull request merged**
